network/src/topology.rs

```diff
@@ -157,6 +172,26 @@ impl KnownNetwork {
                 connections_g.replace(new_connection);
             }
         }
+
+        // Scope the write lock.
+        {
+            let mut nodes_g = self.nodes.write();
+
+            // Remove the nodes that no longer correspond to connections.
+            let node_addrs: HashSet<SocketAddr> = nodes_g.keys().copied().collect();
+            let diff: HashSet<SocketAddr> = node_addrs.difference(&self.nodes_from_connections()).copied().collect();
+            nodes_g.retain(|addr, _| !diff.contains(addr));
```

Comment:
wouldn't it be simpler to retain the nodes that belong to the result of `self.nodes_from_connections()`, i.e. to not use `difference`?

Reply:
yep 👍

Reply:
Why make it simple when you can make it complicated? 🙃